File: common-util-code/src/main/java/com/cjf/ui/text/ExpandTextView.java
package com.cjf.ui.text;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.os.Build;
import android.text.Layout;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.StaticLayout;
import android.text.method.LinkMovementMethod;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.ColorInt;
import androidx.annotation.ColorRes;
import androidx.appcompat.widget.AppCompatTextView;

import com.cjf.util.R;
import com.cjf.util.utils.ResUtils;

/**
 * <p>Title: ExpandTextView </p>
 * <p>Description:  </p>
 * <p>Company: www.mapuni.com </p>
 *
 * @author : 蔡俊峰
 * @version : 1.0
 * @date : 2020/9/10 12:02
 */
public class ExpandTextView extends AppCompatTextView {

    private String originText;// 原始内容文本
    private int initWidth = 0;// TextView可展示宽度
    private int mMaxLines = 3;// TextView最大行数
    private SpannableString SPAN_CLOSE = null;// 收起的文案(颜色处理)
    private SpannableString SPAN_EXPAND = null;// 展开的文案(颜色处理)
    private String mTextExpand = "  展开";
    @ColorRes
    private int mTextExpandColor = R.color.colorAccent;
    private String mTextClose = "  收起";
    private int mTextCloseColor = R.color.colorAccent;

    public ExpandTextView(Context context) {
        this(context, null);
    }

    public ExpandTextView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ExpandTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs, defStyleAttr);
    }

    private void init(Context context, AttributeSet attrs, int defStyleAttr) {
        final TypedArray array = context.obtainStyledAttributes(attrs, R.styleable.ExpandTextView, defStyleAttr, 0);
        try {
            if (array.hasValue(R.styleable.ExpandTextView_textClose)) {
                mTextClose = array.getString(R.styleable.ExpandTextView_textClose);
            }
            if (array.hasValue(R.styleable.ExpandTextView_textCloseColor)) {
                mTextCloseColor = array.getInt(R.styleable.ExpandTextView_textCloseColor, R.color.colorAccent);
            }
            if (array.hasValue(R.styleable.ExpandTextView_textExpandColor)) {
                mTextExpandColor = array.getInt(R.styleable.ExpandTextView_textExpandColor, R.color.colorAccent);
            }
            if (array.hasValue(R.styleable.ExpandTextView_textExpand)) {
                mTextExpand = array.getString(R.styleable.ExpandTextView_textExpand);
            }
        } finally {
            array.recycle();
        }
        initCloseEnd();
    }

    /**
     * 设置TextView可显示的最大行数
     *
     * @param maxLines 最大行数
     */
    @Override
    public void setMaxLines(int maxLines) {
        this.mMaxLines = maxLines;
        super.setMaxLines(maxLines);
    }

    /**
     * 初始化TextView的可展示宽度
     *
     * @param width
     */
    public void initWidth(int width) {
        initWidth = width;
    }

    /**
     * 收起的文案(颜色处理)初始化
     */
    private void initCloseEnd() {
        String content = mTextExpand;
        SPAN_CLOSE = new SpannableString(content);
        ButtonSpan span = new ButtonSpan(getContext(), new OnClickListener() {
            @Override
            public void onClick(View v) {
                ExpandTextView.super.setMaxLines(Integer.MAX_VALUE);
                setExpandText(originText);
            }
        }, mTextCloseColor);
        SPAN_CLOSE.setSpan(span, 0, content.length(), Spanned.SPAN_INCLUSIVE_EXCLUSIVE);
    }

    /**
     * 展开的文案(颜色处理)初始化
     */
    private void initExpandEnd() {
        String content = mTextClose;
        SPAN_EXPAND = new SpannableString(content);
        ButtonSpan span = new ButtonSpan(getContext(), new OnClickListener() {
            @Override
            public void onClick(View v) {
                ExpandTextView.super.setMaxLines(mMaxLines);
                setCloseText(originText);
            }
        }, mTextExpandColor);
        SPAN_EXPAND.setSpan(span, 0, content.length(), Spanned.SPAN_INCLUSIVE_EXCLUSIVE);
    }

    public void setCloseText(CharSequence text) {

        if (SPAN_CLOSE == null) {
            initCloseEnd();
        }
        boolean appendShowAll = false;// true 不需要展开收起功能， false 需要展开收起功能
        originText = text.toString();

        // SDK >= 16 可以直接从xml属性获取最大行数
        int maxLines = 0;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
            maxLines = getMaxLines();
        } else {
            maxLines = mMaxLines;
        }
        String workingText = new StringBuilder(originText).toString();
        if (maxLines != -1) {
            Layout layout = createWorkingLayout(workingText);
            if (layout.getLineCount() > maxLines) {
                //获取一行显示字符个数，然后截取字符串数
                workingText = originText.substring(0, layout.getLineEnd(maxLines - 1)).trim();// 收起状态原始文本截取展示的部分
                String showText = originText.substring(0, layout.getLineEnd(maxLines - 1)).trim() + "..." + SPAN_CLOSE;
                Layout layout2 = createWorkingLayout(showText);
                // 对workingText进行-1截取，直到展示行数==最大行数，并且添加 SPAN_CLOSE 后刚好占满最后一行
                while (layout2.getLineCount() > maxLines) {
                    int lastSpace = workingText.length() - 1;
                    if (lastSpace == -1) {
                        break;
                    }
                    workingText = workingText.substring(0, lastSpace);
                    layout2 = createWorkingLayout(workingText + "..." + SPAN_CLOSE);
                }
                appendShowAll = true;
                workingText = workingText + "...";
            }
        }

        setText(workingText);
        if (appendShowAll) {
            // 必须使用append，不能在上面使用+连接，否则spannable会无效
            append(SPAN_CLOSE);
            setMovementMethod(LinkMovementMethod.getInstance());
        }
    }

    @SuppressLint("SetTextI18n")
    public void setExpandText(String text) {
        if (SPAN_EXPAND == null) {
            initExpandEnd();
        }
        Layout layout1 = createWorkingLayout(text);
        Layout layout2 = createWorkingLayout(text + mTextClose);
        // 展示全部原始内容时 如果 TEXT_CLOSE 需要换行才能显示完整，则直接将TEXT_CLOSE展示在下一行
        if (layout2.getLineCount() > layout1.getLineCount()) {
            setText(originText + "\n");
        } else {
            setText(originText);
        }
        append(SPAN_EXPAND);
        setMovementMethod(LinkMovementMethod.getInstance());
    }

    //返回textview的显示区域的layout，该textview的layout并不会显示出来，只是用其宽度来比较要显示的文字是否过长
    private Layout createWorkingLayout(String workingText) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
            return new StaticLayout(workingText, getPaint(), initWidth - getPaddingLeft() - getPaddingRight(),
                                    Layout.Alignment.ALIGN_NORMAL, getLineSpacingMultiplier(), getLineSpacingExtra(),
                                    false);
        } else {
            return new StaticLayout(workingText, getPaint(), initWidth - getPaddingLeft() - getPaddingRight(),
                                    Layout.Alignment.ALIGN_NORMAL, 1.0f, 0.0f, false);
        }
    }
}

File: common-util-code/src/main/java/com/cjf/ui/text/ExpandTextView.kt
package com.cjf.ui.text

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Color
import android.graphics.Rect
import android.os.Build
import android.text.*
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.text.style.ForegroundColorSpan
import android.text.style.UnderlineSpan
import android.util.AttributeSet
import android.view.View
import androidx.appcompat.widget.AppCompatTextView
import com.cjf.util.R
import com.cjf.util.UtilX
import com.cjf.util.utils.ResUtils

/**
 * <p>Title: ExpandTextView </p>
 * <p>Description:  </p>
 * <p>Company: www.mapuni.com </p>
 *
 * @author : 蔡俊峰
 * @date : 2020/9/18 9:51
 * @version : 1.0
 */
class ExpandTextView : AppCompatTextView {
    /**展开状态 true：展开，false：收起 */
    var expandState: Boolean = false

    /** 状态接口 */
    var mCallback: ExpandCallback? = null

    /** 源文字内容 */
    var mText: CharSequence? = ""

    /** 源文字所有内容 */
    var mTextAll: String? = ""

    /** 开头文字内容 */
    var mTextStart: CharSequence? = ""

    /** 最多展示的行数 */
    var maxLineCount = 3

    /** 省略文字 */
    var ellipsizeText = "..."

    /** 展开文案文字 */
    var expandText: CharSequence = "[全文]"

    /** 展开文案文字颜色 */
    var expandTextColor: Int = Color.parseColor("#1C7FFD")

    /** 收起文案文字 */
    var collapseText = "[收起]"

    /** 收起文案文字颜色 */
    var collapseTextColor: Int = Color.parseColor("#1C7FFD")

    /** 开始文字颜色 */
    var textStartColor: Int = ResUtils.getColor(R.color.design_txt_gray)

    /**是否支持收起功能*/
    var collapseEnable = true

    /** 是否添加下划线 */
    var underlineEnable = false

    constructor(context: Context) : this(context, null) {
    }

    constructor(context: Context, attributes: AttributeSet?) : this(context, attributes, 0) {
    }

    constructor(context: Context, attributes: AttributeSet?, defStyleAttr: Int) : super(context, attributes, defStyleAttr) {
        initTextView(context, attributes, defStyleAttr)
    }

    private fun initTextView(context: Context, attributes: AttributeSet?, defStyleAttr: Int) {
        movementMethod = LinkMovementMethod.getInstance()
        val array = context.obtainStyledAttributes(attributes, R.styleable.ExpandTextView, defStyleAttr, 0)
        try {
            if (array.hasValue(R.styleable.ExpandTextView_textExpand)) {
                expandText = array.getString(R.styleable.ExpandTextView_textExpand) ?: "[全文]"
            }
            if (array.hasValue(R.styleable.ExpandTextView_textCollapse)) {
                collapseText = array.getString(R.styleable.ExpandTextView_textCollapse) ?: "[收起]"
            }
            textStartColor = array.getColor(R.styleable.ExpandTextView_textStartColor, ResUtils.getColor(R.color.design_txt_gray))
            expandTextColor = array.getColor(R.styleable.ExpandTextView_textExpandColor, UtilX.getPrimaryColor())
            collapseTextColor = array.getColor(R.styleable.ExpandTextView_textCollapseColor, UtilX.getPrimaryColor())
            if (array.hasValue(R.styleable.ExpandTextView_maxLineCount)) {
                maxLineCount = array.getInt(R.styleable.ExpandTextView_maxLineCount, 3)
            }
        } finally {
            array.recycle()
        }
    }

    @SuppressLint("DrawAllocation")
    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        // 文字计算辅助工具
        if (mTextAll.isNullOrEmpty()) {
            setMeasuredDimension(measuredWidth, measuredHeight)
        }
        //StaticLayout对象
        val sl = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            StaticLayout.Builder.obtain(
                    mTextAll ?: "",
                    0,
                    mTextAll?.length ?: 0,
                    paint,
                    measuredWidth - paddingLeft - paddingRight
            ).apply {
                setAlignment(Layout.Alignment.ALIGN_CENTER)
            }.build()
        } else {
            StaticLayout(
                    mTextAll, paint, measuredWidth - paddingLeft - paddingRight,
                    Layout.Alignment.ALIGN_CENTER, 1f, 0f,
                    true
            )
        }

        // 总计行数
        var lineCount = sl.lineCount
        //总行数大于最大行数
        if (lineCount > maxLineCount) {
            if (expandState) {
                text = mTextAll
                //是否支持收起功能
                if (collapseEnable) {
                    // 收起文案和源文字组成的新的文字
                    val newEndLineText = mTextAll + collapseText
                    //收起文案和源文字组成的新的文字
                    val spannableString = SpannableStringBuilder(newEndLineText)
                            .apply {
                                //给开始文字设成蓝色
                                if (mTextStart != null) {
                                    setSpan(
                                            ForegroundColorSpan(textStartColor),
                                            0,
                                            mTextAll?.indexOf(mTextStart!!.toString())!! + mTextStart!!.length,
                                            Spannable.SPAN_INCLUSIVE_EXCLUSIVE
                                    )
                                }
                                //给收起设成监听
                                setSpan(
                                        object : ClickableSpan() {
                                            override fun onClick(widget: View) {
                                                if (mCallback != null) {
                                                    mCallback!!.onCollapseClick()
                                                }
                                            }

                                            override fun updateDrawState(ds: TextPaint) {
                                                ds.color = collapseTextColor
                                                ds.isUnderlineText = false
                                            }
                                        },
                                        newEndLineText.length - collapseText.length,
                                        newEndLineText.length,
                                        Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                                )
                                if (underlineEnable) {
                                    //给收起添加下划线
                                    setSpan(
                                            UnderlineSpan(),
                                            newEndLineText.length - collapseText.length,
                                            newEndLineText.length,
                                            Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                                    )
                                }
                            }
                    text = spannableString
                }
                mCallback?.onExpand()
            } else {
                lineCount = maxLineCount
                // 省略文字和展开文案的宽度
                val dotWidth = paint.measureText(ellipsizeText + expandText)
                // 找出显示最后一行的文字
                val start = sl.getLineStart(lineCount - 1)
                val end = sl.getLineEnd(lineCount - 1)
                val lineText = mTextAll?.substring(start, end) ?: ""
                // 将第最后一行最后的文字替换为 ellipsizeText和expandText
                var endIndex = 0
                for (i in lineText.length - 1 downTo 0) {
                    val str = lineText.substring(i, lineText.length)
                    // 找出文字宽度大于 ellipsizeText 的字符
                    if (paint.measureText(str) >= dotWidth) {
                        endIndex = i
                        break
                    }
                }
                // 新的文字
                val newEndLineText = (mTextAll?.substring(0, start) ?: "") + lineText.substring(
                        0,
                        endIndex
                ) + ellipsizeText + expandText
                //全部文字
                val spannableString = SpannableStringBuilder(newEndLineText).apply {
                    //给查看全部设成监听
                    //给开始文字设成蓝色
                    if (mTextStart != null) {
                        setSpan(
                                ForegroundColorSpan(textStartColor),
                                0,
                                mTextAll?.indexOf(mTextStart!!.toString())!! + mTextStart!!.length,
                                Spannable.SPAN_INCLUSIVE_EXCLUSIVE
                        )
                    }
                    setSpan(
                            object : ClickableSpan() {
                                override fun onClick(widget: View) {
                                    if (mCallback != null) {
                                        mCallback!!.onExpandClick()
                                    }
                                }

                                override fun updateDrawState(ds: TextPaint) {
                                    ds.color = expandTextColor
                                    ds.isUnderlineText = false
                                }
                            },
                            newEndLineText.length - expandText.length,
                            newEndLineText.length,
                            Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                    )
                    if (underlineEnable) {
                        setSpan(
                                UnderlineSpan(),
                                newEndLineText.length - expandText.length,
                                newEndLineText.length,
                                Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                        )
                    }
                }
                // 最终显示的文字
                text = spannableString
                mCallback?.onCollapse()
            }
        } else {
            //收起文案和源文字组成的新的文字
            val spannableString = SpannableStringBuilder(mTextAll)
                    .apply {
                        //给开始文字设成蓝色
                        if (mTextStart != null) {
                            setSpan(
                                    ForegroundColorSpan(textStartColor),
                                    0,
                                    mTextAll?.indexOf(mTextStart!!.toString())!! + mTextStart!!.length,
                                    Spannable.SPAN_INCLUSIVE_EXCLUSIVE
                            )
                        }
                    }
            text = spannableString
            mCallback?.onLoss()

        }
        // 重新计算高度
        var lineHeight = 0
        for (i in 0 until lineCount) {
            val lineBound = Rect()
            sl.getLineBounds(i, lineBound)
            lineHeight += lineBound.height()
        }
        lineHeight = (paddingTop + paddingBottom + lineHeight * lineSpacingMultiplier).toInt()
        setMeasuredDimension(measuredWidth, lineHeight)
    }

    /**
     * 设置要显示的文字以及状态
     * @param text
     * @param expanded true：展开，false：收起
     * @param callback
     */
    fun setContentText(textStart: CharSequence, text: CharSequence, expanded: Boolean = false, callback: ExpandCallback = DefaultExpandCallback(this)) {
        mTextStart = textStart;
        mText = text
        mTextAll = textStart.toString() + text.toString()
        expandState = expanded
        mCallback = callback
        // 设置要显示的文字，这一行必须要，否则 onMeasure 宽度测量不正确
        setText(mTextAll)
    }

    /**
     * 设置要显示的文字以及状态
     * @param text
     * @param expanded true：展开，false：收起
     * @param callback
     */
    fun setContentText(text: CharSequence, expanded: Boolean = false, callback: ExpandCallback = DefaultExpandCallback(this)) {
        setContentText("", text, expanded, callback)
    }

    /**
     * 展开收起状态变化
     * @param expanded
     */
    fun setChanged(expanded: Boolean) {
        expandState = expanded
        requestLayout()
    }

    interface ExpandCallback {
        /**
         * 展开状态
         */
        fun onExpand()

        /**
         * 收起状态
         */
        fun onCollapse()

        /**
         * 行数小于最小行数，不满足展开或者收起条件
         */
        fun onLoss()

        /**
         * 点击全文
         */
        fun onExpandClick()

        /**
         * 点击收起
         */
        fun onCollapseClick()
    }

    open class DefaultExpandCallback(private val textView: ExpandTextView) : ExpandCallback {

        override fun onExpand() {

        }

        override fun onCollapse() {
        }

        override fun onLoss() {
        }

        override fun onExpandClick() {
            textView.setChanged(true)
        }

        override fun onCollapseClick() {
            textView.setChanged(false)
        }

    }
}